File: 049_group_anagrams/Solution1/Solution.java
import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, ArrayList<String>> map = new HashMap<>();
        for (String s : strs) {
            char[] charArr = s.toCharArray();
            Arrays.sort(charArr);
            String key = String.valueOf(charArr);
            if (!map.containsKey(key)) map.put(key, new ArrayList<>());
            map.get(key).add(s);
        }

        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        String[] strs = new String[]{"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> results = new Solution().groupAnagrams(strs);
        for (List<String> result : results) {
            for (String tmp : result) {
                System.out.print(tmp + " ");
            }
            System.out.println();
        }
    }
}

File: 049_group_anagrams/Solution1/Solution.py
class Solution:
    def groupAnagrams(self, strs: list) -> list:
        results = {}

        for s in strs:
            key = ''.join(sorted(s))
            if key in results:
                results[key].append(s)
            else:
                results[key] = [s]

        return list(results.values())


if __name__ == '__main__':
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    solution = Solution()
    print(solution.groupAnagrams(strs))

